internal/appsec/emitter/waf/metrics.go

```diff
@@ -0,0 +1,288 @@
+			waf.ContainerTooLarge: telemetry.Count(telemetry.NamespaceAppSec, "waf.input_truncated", []string{"truncation_reason:" + strconv.Itoa(int(waf.ContainerTooLarge))}),
+			waf.ObjectTooDeep:     telemetry.Count(telemetry.NamespaceAppSec, "waf.input_truncated", []string{"truncation_reason:" + strconv.Itoa(int(waf.ObjectTooDeep))}),
+		},
+		truncationDistributions: map[waf.TruncationReason]telemetry.MetricHandle{
+			waf.StringTooLong:     telemetry.Distribution(telemetry.NamespaceAppSec, "waf.truncated_value_size", []string{"truncation_reason:" + strconv.Itoa(int(waf.StringTooLong))}),
+			waf.ContainerTooLarge: telemetry.Distribution(telemetry.NamespaceAppSec, "waf.truncated_value_size", []string{"truncation_reason:" + strconv.Itoa(int(waf.ContainerTooLarge))}),
+			waf.ObjectTooDeep:     telemetry.Distribution(telemetry.NamespaceAppSec, "waf.truncated_value_size", []string{"truncation_reason:" + strconv.Itoa(int(waf.ObjectTooDeep))}),
+		},
+		wafTimerDistributions: map[string]telemetry.MetricHandle{
+			"rasp.duration":     telemetry.Distribution(telemetry.NamespaceAppSec, "rasp.duration", baseTags),
+			"rasp.duration_ext": telemetry.Distribution(telemetry.NamespaceAppSec, "rasp.duration_ext", baseTags),
+			"waf.duration":      telemetry.Distribution(telemetry.NamespaceAppSec, "waf.duration", baseTags),
+			"waf.duration_ext":  telemetry.Distribution(telemetry.NamespaceAppSec, "waf.duration_ext", baseTags),
+		},
+		raspTimeoutCount:  telemetry.Count(telemetry.NamespaceAppSec, "rasp.timeout", baseTags),
+		wafRequestsCounts: xsync.NewMapOf[RequestMilestones, telemetry.MetricHandle](xsync.WithGrowOnly(), xsync.WithPresize(2^5)),
+		raspRuleEval:      raspRuleEval,
+	}
+}
+
+func (m *HandleMetrics) NewContextMetrics() *ContextMetrics {
+	return &ContextMetrics{
+		HandleMetrics: m,
+	}
+}
+
+type ContextMetrics struct {
+	*HandleMetrics
+
+	// SumRASPCalls is the sum of all the RASP calls made by the WAF whatever the rasp rule type it is.
+	SumRASPCalls atomic.Uint32
+	// SumWAFErrors is the sum of all the WAF errors that happened not in the RASP scope.
+	SumWAFErrors atomic.Uint32
+	// SumRASPErrors is the sum of all the RASP errors that happened in the RASP scope.
+	SumRASPErrors atomic.Uint32
+
+	// Milestones are the tags of the metric `waf.requests` that will be submitted at the end of the waf context
+	Milestones RequestMilestones
+}
+
+// RegisterStats increment the metrics for the WAF run stats at the end of each waf context lifecycle
+func (m *ContextMetrics) RegisterStats(stats waf.Stats) {
+	// Add metrics `{waf,rasp}.duration[_ext]`
+	for key, value := range stats.Timers {
+		metric, found := m.wafTimerDistributions[key]
+		if !found {
+			continue
+		}
+
+		// The metrics should be in microseconds
+		metric.Submit(float64(value.Nanoseconds()) / float64(time.Microsecond))
```

Comment:
This is blindly assuming `time.Duration` is nanos (which is true, but...)

```suggestion
		metric.Submit(float64(value.Nanoseconds()) / float64(time.Microsecond.Nanosecods()))
```